### PR TITLE
CHE-9428: fix kubernetes infra after a merge of not checked PR

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?
Fix kubernetes infra after a merge of not checked PR.
Set correct bindings of generics, fix generics;
Add missing properties in che.properties and kubectl deployment.

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by garagatyi** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 1]
```diff
@@ -335,6 +335,12 @@ che.infra.kubernetes.password=
 che.infra.kubernetes.oauth_token=
 che.infra.kubernetes.trust_certs=
 
+# Defines the way how servers are exposed to the world in k8s infra.
+# List of  strategies implemented in Che: default-host, multi-host, single-host
+che.infra.kubernetes.server_strategy=default-host
+# Used to generate domain for a server in a workspace
+che.infra.kubernetes.ingress.domain=unspecified-ingress-domain
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Is it used for each of server strategies?

**Review by akorneta — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 1 above)
Comment (l0rd) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: And what are the possible values?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 2]
```diff
@@ -335,6 +335,12 @@ che.infra.kubernetes.password=
 che.infra.kubernetes.oauth_token=
 che.infra.kubernetes.trust_certs=
 
+# Defines the way how servers are exposed to the world in k8s infra.
+# List of  strategies implemented in Che: default-host, multi-host, single-host
+che.infra.kubernetes.server_strategy=default-host
```
Comment (l0rd) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:340]: How is this related with `che.single.port`? Does this property makes sense if single port is set to `false`?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInfraModule.java [hunk 3]
```diff
@@ -77,8 +78,11 @@ protected void configure() {
     volumesStrategies.addBinding(UNIQUE_STRATEGY).to(UniqueWorkspacePVCStrategy.class);
     bind(WorkspaceVolumesStrategy.class).toProvider(WorkspaceVolumeStrategyProvider.class);
 
-    MapBinder<String, ExternalServerExposerStrategy> ingressStrategies =
-        MapBinder.newMapBinder(binder(), String.class, ExternalServerExposerStrategy.class);
+    MapBinder<String, ExternalServerExposerStrategy<KubernetesEnvironment>> ingressStrategies =
+        MapBinder.newMapBinder(
+            binder(),
+            new TypeLiteral<String>() {},
```
Comment (l0rd) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInfraModule.java:84]: I am curious: do you really need to use `TypeLiteral` for String (that is not a generic type)?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 1 above)
Comment (garagatyi) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: > Is it used for each of server strategies?

no, it is used for `multi-host` strategy only

> And what are the possible values?

Something like `192.168.99.101.nip.io`. Really depends on DNS provider, public multi-level DNS, IP, etc.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 2 above)
Comment (garagatyi) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:340]: It is not related to a single port AFAIK

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInfraModule.java (hunk 3 above)
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInfraModule.java:84]: I need TypeLiteral for `ExternalServerExposerStrategy<KubernetesEnvironment>` and there is no binding method with mixed class and type literal arguments

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 1 above)
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: > no, it is used for multi-host strategy only

Then maybe it makes sense to make it more clear in doc or even property name?
Is it possible that default value `unspecified-ingress-domain` will work? If not, maybe it'd better to leave it empty and fail tomcat start if it is not configured?

Comment (garagatyi) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: I didn't add this property in code, so not sure whether I should change the name now. But I can add a comment about that.
In a case when multi-host is not chosen strategy failure on start when the value is not set would break Che in different valid cases. Value `unspecified-ingress-domain` won't work but hints a user that he needs to address this issue.

Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: > In a case when multi-host is not chosen strategy failure on start when the value is not set would break Che in different valid cases.

I mean failing start only if multi-host is chosen and `che.infra.kubernetes.ingress.domain` is not configured

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInfraModule.java (hunk 3 above)
Comment (l0rd) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInfraModule.java:84]: Ok makes sense thanks

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 2 above)
Comment (l0rd) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:340]: Ok just wondering how `multi-host` works with single port set to `false`. But in any case this is not related to this PR so ok to merge for me.

**Review by l0rd — APPROVED** [review]

**Review by sleshchenko — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 2 above)
Comment (garagatyi) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:340]: AFAIK Single host mode is used by docker infrastructure only. So neither k8s nor openshift installations of Che use it
